lib/Sema/CSBindings.cpp

```diff
@@ -869,6 +869,8 @@ bool TypeVariableBinding::attempt(ConstraintSystem &cs) const {
 
     type = cs.openUnboundGenericType(UGT, locator);
     type = type->reconstituteSugar(/*recursive=*/false);
+  } else if (type->is<InOutType>() && !TypeVar->getImpl().canBindToInOut()) {
```

Comment:
This seems like too late for this kind of a check, probably should be in `getPotentialBindings` somewhere or in `addPotentialBinding`, also I think it should check if type variable allows `l-value` before trying to convert...